Adapt crates/qasm3 to work with recent versions of openqasm3_parser

jlapeyre opened the pull request:
This commit adds no new features or capabilities to the importer. But it brings the importer up to date with the latest version of openqasm3_parser as a preliminary step in improving the importer.

The biggest change in openqasm3_parser is in handling stdgates.inc.  Upon encountering `include "stdgates.inc"` openqasm3_parser reads no file, but rather adds symbols to the symbol table for gates in the standard library.

The function `convert_asg` in the importer calls `oq3_semantics::symbols::SymbolTable.gates` which returns a `Vec` of information about each "declared" gate. The information is the gate's name and signature and SymbolID, which is sufficient to do everything the importer could do before this commit.

Encountering `Stmt::GateDefinition` now is a no-op rather than an error.  (This was previously called `Stmt::GateDeclaration`, but importantly it is really a definition.)

How the standard library, and gates in general, are handled will continue to evolve.

A behavioral difference between this commit and its parent: Before if the importer encountered a gate call before the corresponding definition an error would be raised. With the current commit, the importer behaves as if all gate definitions were moved to the top of the OQ3 program. However, the error will still be found by the parser so that the importer never will begin processing statements.

The importer depends on a particular branch of a copy of openqasm3_parser. When the commit is ready to merge, we will release a version of openqasm3_parser and depend instead on that release.

See https://github.com/openqasm/openqasm/pull/517

jlapeyre committed:
Adapt crates/qasm3 to work with recent versions of openqasm3_parser

This commit adds no new features or capabilities to the importer. But it brings
the importer up to date with the latest version of openqasm3_parser as a preliminary
step in improving the importer.

The biggest change in openqasm3_parser is in handling stdgates.inc.  Upon encountering
`include "stdgates.inc" openqasm3_parser reads no file, but rather adds symbols to the
symbol table for gates in the standard library.

The function `convert_asg` in the importer calls oq3_semantics::symbols::SymbolTable.gates
which returns a `Vec` of information about each "declared" gate. The information is the
gate's name and signature and SymbolID, which is sufficient to do everything the importer
could do before this commit.

Encountering `Stmt::GateDefinition` now is a no-op rather than an error.  (This was
previously called `Stmt::GateDeclaration`, but importantly it is really a definition.)

How the standard library, and gates in general, are handled will continue to evolve.

A behavioral difference between this commit and its parent: Before if the importer
encountered a gate call before the corresponding definition an error would be raised. With
the current commit, the importer behaves as if all gate definitions were moved to the top
of the OQ3 program. However, the error will still be found by the parser so that the
importer never will begin processing statements.

The importer depends on a particular branch of a copy of openqasm3_parser. When
the commit is ready to merge, we will release a version of openqasm3_parser and
depend instead on that release.

See https://github.com/openqasm/openqasm/pull/517

mtreinish commented on crates/qasm3/Cargo.toml:
```diff
@@ -18,4 +18,4 @@ extension-module = ["pyo3/extension-module"]
 indexmap.workspace = true
 hashbrown.workspace = true
-oq3_semantics = "0.0.7"
+oq3_semantics = { git = "https://github.com/jlapeyre/openqasm3_parser", branch = "update-for-qiskit-importer"}
```
Is there a release on crates.io that we should be using for this? I don't think we want to pull in a branch from git to build the parser?

jlapeyre replied:
It was my intention to depend on a released version before merging this PR.

jlapeyre replied:
I opened this PR on March 26. I tagged and released parser v 0.5.0 on March 25. I'm pretty sure that was the intended version. I'll check.

jlapeyre replied:
I published oq3_semantics (and all crates) 0.6.0. And edited the Qiskit Cargo.toml to depend on it. fd2c6d5

jakelishman commented on crates/qasm3/src/build.rs:
```diff
@@ -226,42 +226,37 @@ impl BuilderState {
+    // Gates mapped via CustomGates will not raise an exception.
+    fn map_gate_ids(&mut self, _py: Python, ast_symbols: &SymbolTable) -> PyResult<()> {
+        for (name, name_id, defined_num_params, defined_num_qubits) in ast_symbols.gates().iter() {
+            let name = name.to_string();
```
Does this need converting to an owned type?  I don't see anything that needs an allocation.

jlapeyre replied:
Fixed in f7cf821

jakelishman commented on crates/qasm3/src/build.rs:
```diff
@@ -226,42 +226,37 @@ impl BuilderState {
+            if name == "U" {
+                // The sole built in gate. `gphase` is treated specially.
+                continue;
+            }
```
Can we arrange the symbol table to make this check unnecessary?  I'd expect all the built-in symbols to have been populated into the symbol table, so there's no special-case behaviour needed to handle them for a consumer; we could just fetch the symbol and get a gate.

jlapeyre replied:
Are you thinking for example that `SymbolTable::gates` does not return `"U"` in its list and this check is omitted?
In the parser, it's convenient to make `U` a symbol of type `Gate`.  For example, we don't need to handle checking number and type of args specially when it's called.

jlapeyre replied:
Fixed (made the change in my previous comment) in b22b7d7

jakelishman replied:
Maybe I misunderstood why this check was here - I'm expecting `U` to be both in `SymbolTable::gates` _and_ to be handled in the same way as other gates in this function.

jakelishman commented on crates/qasm3/src/build.rs:
```diff
@@ -377,37 +372,45 @@ pub fn convert_asg(
     };
+
+    let _ = state.map_gate_ids(py, ast_symbols);
+
```
This is explicitly ignoring a `Result`, with the one bit of syntax that gets around the clippy lint for not "using" a `Result` return value.  I think what's intended is `state.map_gate_ids(py, ast_symbols)?;` with no `let`.   It'd be good to add a test that the exceptions are raised in Python space - we didn't do it at the initial merge because of a rush to land the PR, but we ought to back fill.

jlapeyre replied:
I thought clippy complains about throwing away *any* return value. I agree that `state.map_gate_ids(py, ast_symbols)?` is much better.

jlapeyre replied:
Partly fixed in 258499c
But I did not yet write a new test.

jakelishman replied:
`Result` as a type is marked `#[must_use]`, so any expression that evaluates to a `Result` needs to be "used" in some form.  Most types don't have that annotation, though some functions returning normally safe types might have it, so you may have seen it there.

jakelishman commented on crates/qasm3/src/build.rs:
```diff
@@ -377,37 +372,45 @@ pub fn convert_asg(
             asg::Stmt::DeclareClassical(decl) => state.declare_classical(py, ast_symbols, decl)?,
             asg::Stmt::DeclareQuantum(decl) => state.declare_quantum(py, ast_symbols, decl)?,
-            asg::Stmt::GateDeclaration(decl) => state.define_gate(py, ast_symbols, decl)?,
+            // We ignore gate definitions because the only information we can currently use
+            // from them is extracted with `SymbolTable::gates` via `map_gate_ids`.
+            asg::Stmt::GateDefinition(_) => (),
```
Doing this is fine, but leads me to check: does the backing parser correctly handle the order of definitions?  For example, will it correctly reject the program
```qasm
qubit q;
my_gate q;
gate my_gate q {}
```
because `my_gate` isn't defined at the point of use?

I'm guessing so - I'm just checking because this swap from a single-pass iteration in the Qiskit side to relying on the AST means that the responsibility for handling it moves away from us.

jlapeyre replied:
Yes it is rejected:
```
Error: UndefGateError
   ╭─[/path/to/examples/qasm/ex1.qasm:2:1]
 2 │my_gate q;
   │   ╰──── Near this point
```
There were some other doubts I had, a bit vague, about this approach. I don't recall what they are. But if they are warranted, it should become clear later.

jakelishman commented on crates/qasm3/src/build.rs:
```diff
@@ -226,42 +226,37 @@ impl BuilderState {
+    // Encountering any gates not in the standard library results in raising an exception.
+    // Gates mapped via CustomGates will not raise an exception.
+    fn map_gate_ids(&mut self, _py: Python, ast_symbols: &SymbolTable) -> PyResult<()> {
+        for (name, name_id, defined_num_params, defined_num_qubits) in ast_symbols.gates().iter() {
```
Assuming `SymbolTable::gates` produces some sort of `&[T]`, the `.iter()` probably isn't necessary because `into_iter()` (which `for` calls implicitly) on `&[T]` consumes the reference but not the `T`, so there's no borrow-check problem.

jlapeyre replied:
[Here is `SymbolTable::gates`](https://github.com/Qiskit/openqasm3_parser/blob/92dead66a0f27d5a12ec301d10293f6e453b0034/crates/oq3_semantics/src/symbols.rs#L262-L276) : 
```rust
    /// Return a Vec of information about all gate declarations. Each element
    /// is a tuple of (gate name, symbol id, num classical params, num quantum params).
    pub fn gates(&self) -> Vec<(&str, SymbolId, usize, usize)> {
```
It ends with `collect()` because I was unable to quickly come up with a trait bound for the return type.... or maybe I never even tried for this one. But I think it is more flexible to return the iterator, right?

jlapeyre replied:
I removed `.iter()`. 
Fixed in d7a5e61

jlapeyre replied:
I also removed the final `collect` in `SymbolTable::gates`. The return type is no longer `Vec`, but is now `impl Iterator<Item = (&str, SymbolId, usize, usize)>`. But this required no further edits to the present PR. I mean d7a5e61 was enough. This change is present in oq3_semantics 0.6.0. The present PR depends on this version.

jakelishman replied:
Ah, that's good to both removing the `iter` and making the function return some unnamed `impl Iterator` thanks.

jakelishman reviewed (commented):
Minor code comments and my agreement with Matt's comment aside, this looks good to me.

jlapeyre committed:
Remove unnecessary conversion `name.to_string()`

Response to reviewer comment https://github.com/Qiskit/qiskit/pull/12087/files#r1586949717

jlapeyre committed:
Remove check for U gate in map_gate_ids

* This requires modifying the external parser crate.
* Depend temporarily on the branch of the parser with this modification.
* Make another change required by other upstream improvments.
* Cargo config files are changed because of above changes.

jlapeyre committed:
Return error returned by map_gate_ids in convert_asg

Previously this error was ignored. This would almost certainly cause
an error later. But better to do it at the correct place.

jlapeyre committed:
Remove superfluous call to `iter()`

jlapeyre commented:
I think all the changes requested have been made. Except one:

No tests yet added. See https://github.com/Qiskit/qiskit/pull/12087#discussion_r1586951469

I think that issues in this comment  https://github.com/Qiskit/qiskit/pull/12087#discussion_r1586953801 have been resolved.

This is done: ~~However, I could make `SymbolTable::gates` return an iterator, and consume it here. On the other hand, there is no additional copy or allocation done here.~~

jlapeyre committed:
Depend on published oq3_semantics 0.6.0

mtreinish committed:
Fix cargo lock file

mtreinish committed:
Merge remote-tracking branch 'origin/main' into update-for-openqasm3_parser-changes

mtreinish reviewed (approved):
This LGTM, I'm not as familiar with the oq3 parser code but nothing looks clearly wrong and this is passing all the tests we have and is using the latest release of `oq3_semantics` which is the goal of the PR. We can continue to iterate on this if any issues are raised during the rc period.
